cryostatio/cryostat-operator
# feat(auth): deploy openshift auth proxy

# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
_______________________________________________

Related to #710
Based on (merged in) #800

## Description of the change:
Deploys `openshift-oauth-proxy` in the Pod and points the existing Service (and Route or Ingress) at it, rather than at Cryostat directly. The auth proxy is always configured to require one particular set of Roles/permissions for both interactive users as well as programmatic clients, and always delegates to OpenShift SSO. In this PR it cannot be disabled and it cannot be augmented with htpasswd Basic auth. It also does not handle TLS.

## Motivation for the change:
In Cryostat 3.0 there should always be an auth proxy placed in front of the operand containers, since in the new architecture the authz responsibility is always delegated to such an auth proxy and never performed by the Cryostat container. The OpenShift OAuth Proxy with SubjectAccessReview and TokenReview configuration in this manner is very close to what Cryostat <= 2.4 did and generally allows users and clients to authenticate and access the application and API with the same permissions as before.

## How to manually test:
1. `OPERATOR_IMG=quay.io/andrewazores/cryostat-operator:3.0.0-openshift-oauth-proxy-3 DISABLE_SERVICE_TLS=true DEPLOY_NAMESPACE=$(oc project -q) make install deploy`
2. Create a Cryostat CR with `enableCertManager: false`
3. Wait for deployment to become ready, then open the route URL in your browser
4. Try to log in with an unprivileged user, ex. `developer`. You should see a "permission denied" page. Click "Log in" again and this time log in with a privileged user, ex. `kubeadmin`. You should get in to the application and be able to perform some basic functionality ex view Security certs/credentials lists (many still broken due to https://github.com/openshift/oauth-proxy/issues/272)
5. Log out and verify that repeating step 4 behaves the same way again
6. In a terminal, use http/curl/wget: `http --follow cryostat-sample-cryostat3-helm.apps-crc.testing/api/v3/tls/certs` produces a login page response, while `http --follow --auth-type=bearer --auth=$(oc whoami -t) cryostat-sample-cryostat3-helm.apps-crc.testing/api/v3/tls/certs` returns the list response (assuming you are logged in with `oc` as a privileged user). `http --follow --auth-type=bearer --auth=FAKE$(oc whoami -t) cryostat-sample-cryostat3-helm.apps-crc.testing/api/v3/tls/certs` (a bad/corrupted token) should result in the login page response.
